src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -694,10 +694,13 @@ int MetaspaceShared::preload_classes(const char* class_list_path, TRAPS) {
     }
     Klass* klass = parser.load_current_class(THREAD);
     if (HAS_PENDING_EXCEPTION) {
-      if (klass == NULL &&
-          (PENDING_EXCEPTION->klass()->name() == vmSymbols::java_lang_ClassNotFoundException())) {
-        // print a warning only when the pending exception is class not found
-        log_warning(cds)("Preload Warning: Cannot find %s", parser.current_class_name());
+      if (klass == NULL) {
+        Symbol* exception_klass_name = PENDING_EXCEPTION->klass()->name();
+        if (exception_klass_name == vmSymbols::java_lang_ClassNotFoundException() ||
+            exception_klass_name == vmSymbols::java_lang_UnsupportedClassVersionError()) {
+          // print a warning only when the pending exception is class not found
+          log_warning(cds)("Preload Warning: Cannot find %s", parser.current_class_name());
+        }
```

Comment:
It seems the other code path which would print the "Preload Warning: Cannot find..." message is as follows:
```
ClassLoaderExt::load_class(Symbol *, const char *, Thread *) : InstanceKlass *
	ClassListParser::load_class_from_source(Symbol *, Thread *) : InstanceKlass *
		ClassListParser::load_current_class(Thread *) : Klass *
			MetaspaceShared::preload_classes(const char *, Thread *) : int
				MetaspaceShared::preload_and_dump(Thread *) : void (2 matches)
```
which is for custom class loaders and your change in classListParser.cpp did not touch.

Anyway, if the added condition is necessary, the comment should be adjusted to mention the condition.

Reply:
Hi Calvin, thanks for checking this. I traced the code and found out what happened:

Before this patch, when loading an old class unsupported by CDS, `JavaCalls::call_virtual()` inside `ClassListParser::load_current_class()` will throw `java.lang.UnsupportedClassVersionError`. This exception is caught and discarded.

The code will then call `SystemDictionary::resolve_or_null()`, which returns `NULL` because the class is not found inside the boot loader. At this point, `load_current_class()` throws `java_lang_ClassNotFoundException`.  `MetaspaceShared::preload_and_dump()` sees this and prints the "Preload Warning: Cannot find..." message.

After this patch, the `UnsupportedClassVersionError` is not caught anymore, so it will be seen inside `MetaspaceShared::preload_and_dump()`.

I updated the comments to mention the condition. I also created a new RFE (https://bugs.openjdk.java.net/browse/JDK-8263469) to improve the warning messages.

Reply:
Thanks for the explanations.